lucet-runtime/lucet-runtime-internals/src/instance.rs

```diff
@@ -800,41 +800,58 @@ impl Instance {
 
         // Sandbox has jumped back to the host process, indicating it has either:
         //
-        // * trapped, or called hostcall_error: state tag changed to something other than `Running`
+        // * returned: state should be `Running`; transition to Ready and return a RunResult
+        // * yielded: state should be `Yielded`; take the yielded value and return a RunResult
+        // * trapped: state should be `Fault`; populate details and return an error or call a hanlder as appropriate
+        // * terminated: state should be `Terminated`; take the termination details and return as an Err
         // * function body returned: set state back to `Ready` with return value
+        //
+        // The state should never be `Ready` at this point
 
-        match &self.state {
+        match &mut self.state {
```

Comment:
(the `mut` here is what got me thinking about it, so the comment is here)

summarizing our conversation: especially now that the C API isn't fiddling with `State` directly, it's not an additional complexity for users to split `State` into more granular variants to distinguish between `Yielded` (returning a result) and `Expecting` (to restart with an indicated value) - similarly for `Terminated` -> `TerminatedWith` (some fault information) and `Terminated` (the "this instance must be restarted to execute"). This would let us avoid the `Option<_>` and `take()` by just taking the enum fields for an eventual `RunResult` and directly resetting `state` to the appropriate instance state.

Parceling `State` into `payload`, `can execute`, and `must be reset` states might be interesting conceptually or for documentation reasons but probably not interesting to work into the enum.

Reply:
Done in 56970cb; thanks for the good suggestion :+1: